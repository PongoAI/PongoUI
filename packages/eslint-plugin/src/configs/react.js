/** @type {import("eslint").Linter.Config} */
const config = {
  extends: ['airbnb-typescript', 'prettier'],
  parser: '@typescript-eslint/parser',
  parserOptions: {
    project: 'tsconfig.json',
  },
  plugins: ['deprecation', 'import', '@typescript-eslint', 'jest', 'jsx-a11y', 'react', 'react-hooks'],
  settings: {
    'import/parsers': {
      '@typescript-eslint/parser': ['.ts', '.tsx'],
    },
    'import/resolver': {
      typescript: {
        alwaysTryTypes: true,
      },
    },
    'no-unused-expressions': 'off',
  },
  env: {
    browser: true,
    es6: true,
    node: true,
  },

  ignorePatterns: ['coverage', 'dist', 'dist-storybook', 'lib', 'lib-commonjs', '**/__snapshots__', 'node_modules'],
  rules: {
    'no-unused-vars': 'error',
    'no-use-before-define': 'off',
    '@typescript-eslint/no-use-before-define': ['error'],
    'import/no-extraneous-dependencies': ['error', { devDependencies: true }],
    'import/extensions': [
      'error',
      'ignorePackages',
      {
        js: 'never',
        jsx: 'never',
        ts: 'never',
        tsx: 'never',
      },
    ],
  },
};

module.exports = config;
