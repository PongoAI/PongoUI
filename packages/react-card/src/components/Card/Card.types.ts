import type { ComponentProps, ComponentState, Slot } from '@fluentui/react-utilities';
import React from 'react';

export type CardSlots = {
  /**
   * The root of the Card.
   */
  root: Slot<'div', 'button'>;
};

export type CardCommons = {
  /**
   * The appearance of the Card component.
   *
   * - elevated: The Card is elevated through drop shadow.
   * - outlined: The Card contains an outline border.
   *
   * @defaultValue elevate
   */
  appearance?: 'elevate' | 'outline';

  /**
   * The shape of the Card.
   *
   * @defaultValue rounded
   */
  shape?: 'rounded' | 'circular' | 'circle' | 'square';

  /**
   * Whether the intractable Card should be disabled.
   *
   * @defaultValue false
   */
  disabled?: boolean;

  /**
   * Whether the Card should be rendered inline.
   *
   * @defaultValue false
   */
  inline?: boolean;

  /**
   * Change event to call when the user interacts with the card (press or keyboard).
   * This is used for intractable Card content.
   */
  onClick?: (ev: React.MouseEvent | React.KeyboardEvent) => void;
};

export interface CardProps extends Omit<ComponentProps<CardSlots>, 'onClick'>, CardCommons {}

export type CardState = ComponentState<CardSlots> & CardCommons;
