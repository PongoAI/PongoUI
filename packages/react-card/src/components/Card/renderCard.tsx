import * as React from 'react';
import { getSlots } from '@fluentui/react-utilities';
import type { CardState, CardSlots } from './Card.types';

export const renderCard = (state: CardState) => {
  const { slots, slotProps } = getSlots<CardSlots>(state);

  return <slots.root {...slotProps.root}>{state.root.children}</slots.root>;
};
