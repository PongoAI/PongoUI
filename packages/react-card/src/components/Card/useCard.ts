import * as React from 'react';
import { getNativeElementProps } from '@fluentui/react-utilities';
import type { CardProps, CardState } from './Card.types';
import { useFocusableGroup } from '@fluentui/react-tabster';

export const useCard = (props: CardProps, ref: React.Ref<HTMLElement>): CardState => {
  const { appearance = 'elevate', inline = false, shape = 'rounded', disabled = false, onClick } = props;
  const groupFocusAttributes = useFocusableGroup({ tabBehavior: 'limitedTrapFocus' });

  const state: CardState = {
    appearance,
    shape,
    disabled,
    inline,
    root: getNativeElementProps(props.as || onClick ? 'button' : 'div', {
      ref,
      role: 'group',
      type: onClick && 'button',
      ...groupFocusAttributes,
      ...props,
    }),
    components: {
      root: onClick ? 'button' : 'div',
    },
  };

  // If the card can be interacted with, forward the callback to the events.
  if (!disabled && onClick) {
    state.root.onClick = (ev: React.MouseEvent<HTMLElement> | React.KeyboardEvent<HTMLElement>) => {
      onClick?.(ev);
    };
  }

  return state;
};
