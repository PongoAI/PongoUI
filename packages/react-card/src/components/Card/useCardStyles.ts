import { makeStyles, mergeClasses, shorthands } from '@griffel/react';
import type { CardState } from './Card.types';
import { tokens } from '@pongo-ui/react-theme';

const useStyles = makeStyles({
  root: {
    flexDirection: 'column',
    backgroundColor: tokens.canvasColor,
    ...shorthands.padding(tokens.focusedLayout),
    flexShrink: 0,
    ...shorthands.gap(tokens.focusedLayout),
  },

  inline: {
    display: 'inline-flex',
  },

  flex: {
    display: 'flex',
  },

  /**
   * Appearance styles
   */
  elevate: {
    filter: tokens.elevate,
  },

  outline: {
    ...shorthands.border('1px', 'solid', tokens.inherit),
  },

  /**
   * Interactive styles
   */
  intractable: {
    cursor: 'pointer',

    ':hover': {
      backgroundColor: tokens.inheritForegroundHover,
    },
    ':active': {
      backgroundColor: tokens.inheritForegroundPressed,
    },
  },

  disabled: {
    cursor: 'not-allowed',
  },

  /**
   * Shape styles
   */
  rounded: {
    ...shorthands.borderRadius(tokens.rounded),
  },

  circular: {
    ...shorthands.borderRadius(tokens.circular),
  },

  circle: {
    ...shorthands.borderRadius(tokens.circle),
  },

  square: {
    ...shorthands.borderRadius(tokens.square),
  },
});

export const useCardStyles = (state: CardState) => {
  const styles = useStyles();

  state.root.className = mergeClasses(
    styles.root,
    styles[state.appearance!],
    styles[state.shape!],
    state.inline ? styles.inline : styles.flex,
    !state.disabled &&
      (state.root.onClick ||
        state.root.onMouseUp ||
        state.root.onMouseDown ||
        state.root.onPointerUp ||
        state.root.onPointerDown ||
        state.root.onTouchStart ||
        state.root.onTouchEnd) &&
      styles.intractable,
    state.disabled && styles.disabled,
    state.root.className,
  );

  return state;
};
