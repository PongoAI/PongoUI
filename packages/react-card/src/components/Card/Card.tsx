import * as React from 'react';
import { useCard } from './useCard';
import { renderCard } from './renderCard';
import { useCardStyles } from './useCardStyles';
import type { CardProps } from './Card.types';
import type { ForwardRefComponent } from '@fluentui/react-utilities';

/**
 * The Card component displays content and presents actions to the user.
 */
export const Card: ForwardRefComponent<CardProps> = React.forwardRef((props, ref) => {
  const state = useCard(props, ref);

  useCardStyles(state);

  return renderCard(state);
});
Card.displayName = 'Card';
