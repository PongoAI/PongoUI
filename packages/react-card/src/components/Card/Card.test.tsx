import * as React from 'react';
import { Card } from './Card';
import { isConformant } from '../../common/isConformant';
import { render } from '@testing-library/react';

describe('Card', () => {
  isConformant({
    Component: Card,
    displayName: 'Card',
  });

  it('renders a basic Card', () => {
    const renderedComponent = render(<Card>This is a card.</Card>);

    expect(renderedComponent.baseElement).toMatchSnapshot();
  });
});
