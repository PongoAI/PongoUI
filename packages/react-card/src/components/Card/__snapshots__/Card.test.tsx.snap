// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Card renders a basic Card 1`] = `
<body>
  <div>
    <div
      class="___1ba5muk f1vx9l62 f8qjfyk fk7qxxv f10sqwl7 f1eg49sl fne32i3 fi64zpg fnmxc1e fynvm4a f1ern45e f1n71otn f1h8hb77 f1deefiw fl4n2cu f1rbvai9 f13ju9t4 f1lh17z1 f1dlw1sp f22iagw"
      data-tabster="{\\"groupper\\":{\\"tabbability\\":2}}"
      role="group"
    >
      This is a card.
    </div>
  </div>
</body>
`;
